services/user/src/types/custom.ts

```diff
@@ -1,8 +1,38 @@
 import { Types } from 'mongoose';
+import { z } from 'zod';
+
+export enum UserValidationErrors {
+    REQUIRED = 'REQUIRED',
+    INVALID = 'INVALID',
+}
 
 export interface RequestUser {
     id: Types.ObjectId | string;
     username: string;
     email: string;
     isAdmin: boolean;
 }
+
+export const userSchema = z.object({
+    username: z
+        .string({
+            invalid_type_error: UserValidationErrors.INVALID,
+            required_error: UserValidationErrors.REQUIRED,
+        })
+        .regex(/^[a-zA-Z0-9._-]+$/, UserValidationErrors.INVALID),
+    email: z
+        .string({
+            invalid_type_error: UserValidationErrors.INVALID,
+            required_error: UserValidationErrors.REQUIRED,
+        })
+        .email(UserValidationErrors.INVALID),
+    password: z
+        .string({
+            invalid_type_error: UserValidationErrors.INVALID,
+            required_error: UserValidationErrors.REQUIRED,
+        })
+        .regex(
+            /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.{8,})(?=.*[!"#$%&'()*+,-.:;<=>?@\\/\\[\]^_`{|}~])/,
+            UserValidationErrors.INVALID,
+        ),
+});
```

Comment:
Should we use multiple regexes to express multiple errors?

Reply:
I was assuming this was just for simple validation and we wouldn't be differentiating between the different errors. Is there a use case you have in mind or is this for readability and maintainability?

Reply:
Ok yea I should not review PRs at 2am I think this is fine